staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -77,4 +78,33 @@ type Response struct {
 	Audiences Audiences
 	// User is the UserInfo associated with the authentication context.
 	User user.Info
+	// AuthMethod identifies the method used to authenticate the request (ex. token or x509).
+	AuthMethod string
+	// AuthenticatorName identifies the authenticator that authenticated the request.
+	AuthenticatorName string
+}
+
+// Error is the error returned by an authenticator upon failures.
+type Error struct {
+	// Err is the error produced by an authenticator.
+	Err error
+	// AuthMethod identifies the method used to authenticate the request (ex. token or x509).
+	AuthMethod string
+	// AuthenticatorName identifies the authenticator that failed an authentication request.
+	AuthenticatorName string
+}
+
+// Error implements the errors.Error interface.
+func (a Error) Error() string {
+	return fmt.Sprintf("authentcation method: %q by authenticator: %q failed to process authentication request: %v", a.AuthMethod, a.AuthenticatorName, a.Err)
+}
+
+// Unwrap unpacks the wrapped error.
+func (a Error) Unwrap() error {
+	return a.Err
+}
+
+// NewError constructs Error.
+func NewError(authMethod, authenticatorName string, err error) error {
+	return &Error{AuthenticatorName: authenticatorName, AuthMethod: authMethod, Err: err}
 }
```

Comment:
This might save some typing.

```
func Errorf(authMethod, authenticatorName, format string, a ...interface{}) error {
	return &Error{AuthenticatorName: authenticatorName, AuthMethod: authMethod, Err: fmt.Errorf(format, a...)}
}
```

Then we don't need to do stuff like

```
protocolAuthError(errors.New(...))
```

Reply:
Added Errorf.
In the case of protocol.go, usign Errorf did not actually simplify things, but it did in other places.
PTAL.